model/result/ScoreWriterService.php

```diff
@@ -0,0 +1,121 @@
+
+/**
+ * Class LtiXmlFormatterService
+ * Class to manage XML result data with score and to store it in DeliveryExecution
+ * @package oat\taoLtiConsumer\model\classes
+ */
+class ScoreWriterService extends ConfigurableService
+{
+    /**
+     * Store the score result into a delivery execution
+     * @param $result
+     * @return string
+     * @throws ResultException
+     * @throws common_exception_Error
+     * @throws DuplicateVariableException
+     */
+    public function store($result)
+    {
+        if (!(isset($result['score']) && $this->isScoreValid($result['score']))) {
+            throw new InvalidScoreException(
+                MessagesService::$statuses[MessagesService::STATUS_INVALID_SCORE], MessagesService::STATUS_INVALID_SCORE, null,
+                MessagesService::buildMessageData(MessagesService::STATUS_INVALID_SCORE, $result)
+            );
+        }
+
+        if (!isset($result['sourcedId'])) {
+            throw new ResultException(MessagesService::$statuses[MessagesService::STATUS_DELIVERY_EXECUTION_NOT_FOUND],
+                MessagesService::STATUS_DELIVERY_EXECUTION_NOT_FOUND, null,
+                MessagesService::buildMessageData(MessagesService::STATUS_DELIVERY_EXECUTION_NOT_FOUND, $result)
+            );
+        }
+
+        $deliveryExecution = $this->getDeliveryExecution($result);
+
+
+        /** @var ResultServerService $resultServerService */
+        $resultServerService = $this->getServiceLocator()->get(ResultServerService::SERVICE_ID);
+        $resultStorageService = $resultServerService->getResultStorage($result['sourcedId']);
+        $resultStorageService->storeTestVariable($result['sourcedId'], '', $this->getScoreVariable($deliveryExecution->getIdentifier(), $result['score']), '');
+
+        return $deliveryExecution->getIdentifier();
+    }
+
+    /**
+     * Look for a DeliveryExecution and return it or throw an Exception
+     * @param array $result
+     * @throws ResultException
+     * @return DeliveryExecution
+     */
+    private function getDeliveryExecution($result)
+    {
+        try {
+            /** @var ServiceProxy $resultService */
+            $resultService = $this->getServiceLocator()->get(ServiceProxy::SERVICE_ID);
+            $deliveryExecution = $resultService->getDeliveryExecution($result['sourcedId']);
+            $deliveryExecution->getDelivery();
+        } catch (\Exception $e) {
+            throw new ResultException($e->getMessage(), MessagesService::STATUS_DELIVERY_EXECUTION_NOT_FOUND, null,
+                MessagesService::buildMessageData(MessagesService::STATUS_DELIVERY_EXECUTION_NOT_FOUND, $result)
+            );
+        }
+
+        return $deliveryExecution;
+    }
+
+    /**
+     * @param $identifier
+     * @param string $score
+     * @return ResultServerOutcomeVariable
+     * @throws common_exception_InvalidArgumentType
+     */
+    private function getScoreVariable($identifier, $score)
```

Comment:
Maybe no need to change now, but just noticing that this could maybe belong to a factory...

Reply:
I have added TODO: Move ResultServerOutcomeVariable creation into a factory